pipeline/config_test.go

```diff
@@ -0,0 +1,263 @@
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser", "json_parser")})
+
+				return ops
+			},
+			func() Config {
+				var ops Config
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser1", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser2", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser3", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser4", "json_parser")})
+				return ops
+			}(),
+		},
+		{
+			"different_ops",
+			func() Config {
+				var ops Config
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("copy", "copy")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("copy", "copy")})
+
+				return ops
+			},
+			func() Config {
+				var ops Config
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser1", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser2", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("copy", "copy")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("copy1", "copy")})
+				return ops
+			}(),
+		},
+		{
+			"unordered",
+			func() Config {
+				var ops Config
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("copy", "copy")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("copy", "copy")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser", "json_parser")})
+
+				return ops
+			},
+			func() Config {
+				var ops Config
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("copy", "copy")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser1", "json_parser")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("copy1", "copy")})
+				ops = append(ops, operator.Config{Builder: newDummyOpConfig("json_parser2", "json_parser")})
+				return ops
+			}(),
+		},
+		{
+			"already_renamed",
```

Comment:
I'd like to see some more variations on this one.

Cascading collisions:
```
- copy
- copy
- copy1
- copy2
- copy2
```


Out of order:
```
- copy2
- copy
- copy
- copy
```

As suggested above, "stealing ids"
```
- copy
- copy
- copy1
```

Can you think of others?

Reply:
won't that first part not even make it into deduplicate function as they have duplicate values?

Reply:
Yep, you're right. The autogen functionality does not dedplicate all IDs, it just autogenerates IDs and makes sure they are not the same as other IDs.

Reply:
I would still like to see a "already renamed but out of order" test:
```
"copy1"
"copy"
"copy"
```